bootstrap/pkg/apis/apps/kfdef/kfloader.go

```diff
@@ -0,0 +1,292 @@
+	}
+}
+
+// Copy GCP plugin spec. Will skip if platform is not GCP.
+func copyGcpPluginSpec(from *kfdefv1alpha1.KfDef, to *kfdefv1beta1.KfDef) error {
+	if from.Spec.Platform != kftypesv3.GCP {
+		return nil
+	}
+
+	spec := kfgcp.GcpPluginSpec{}
+	if err := to.GetPluginSpec(kftypesv3.GCP, &spec); err != nil && !kfdefv1beta1.IsPluginNotFound(err) {
+		return err
+	}
+	spec.Project = from.Spec.Project
+	spec.Email = from.Spec.Email
+	spec.IpName = from.Spec.IpName
+	spec.Hostname = from.Spec.Hostname
+	spec.Zone = from.Spec.Zone
+	spec.UseBasicAuth = from.Spec.UseBasicAuth
+	spec.SkipInitProject = from.Spec.SkipInitProject
+	spec.UseIstio = from.Spec.UseIstio
+	spec.EnableApplications = from.Spec.EnableApplications
+	spec.ServerVersion = from.Spec.ServerVersion
+	spec.DeleteStorage = from.Spec.DeleteStorage
+	return to.SetPluginSpec(kftypesv3.GCP, spec)
+}
+
+// Copy plugins configs.
+func copyPlugins(from *kfdefv1alpha1.KfDef, to *kfdefv1beta1.KfDef) {
+	for _, plugin := range from.Spec.Plugins {
+		betaPlugin := kfdefv1beta1.Plugin{}
+		betaPlugin.Name = plugin.Name
+		betaPlugin.Kind = alphaPluginNameToBetaKind(plugin.Name)
+		if plugin.Spec != nil {
+			betaPlugin.Spec = &runtime.RawExtension{}
+			*betaPlugin.Spec = *plugin.Spec
+		}
+		to.Spec.Plugins = append(to.Spec.Plugins, betaPlugin)
+	}
+
+	specCopiers := []func(*kfdefv1alpha1.KfDef, *kfdefv1beta1.KfDef) error{
+		copyGcpPluginSpec,
+	}
+	for _, copier := range specCopiers {
+		if err := copier(from, to); err != nil {
+			log.Errorf("having error when copying plugin spec: %v", err)
+		}
+	}
+}
+
+// Load v1alpha1 KfDef and returns v1beta1 KfDef.
+func loadKfDefV1Alpha1(configs []byte) (*kfdefv1beta1.KfDef, error) {
+	alphaKfDef := &kfdefv1alpha1.KfDef{}
+	if err := yaml.Unmarshal(configs, alphaKfDef); err != nil {
+		return nil, &kfapis.KfError{
+			Code:    int(kfapis.INVALID_ARGUMENT),
+			Message: fmt.Sprintf("invalid config file format: %v", err),
+		}
+	}
+
+	betaKfDef := &kfdefv1beta1.KfDef{
+		TypeMeta:   alphaKfDef.TypeMeta,
+		ObjectMeta: alphaKfDef.ObjectMeta,
+	}
+	betaKfDef.APIVersion = Api + "/v1beta1"
+
+	// Converting functions wrapper.
+	converters := []func(*kfdefv1alpha1.KfDef, *kfdefv1beta1.KfDef){
+		copyApplications,
+		copyRepos,
+		copySecrets,
+		copyPlugins,
+	}
+	for _, converter := range converters {
+		converter(alphaKfDef, betaKfDef)
+	}
+
+	return betaKfDef, nil
+}
+
+// Load v1beta1 KfDef.
+func loadKfDefV1Beta1(configs []byte) (*kfdefv1beta1.KfDef, error) {
+	kfdef := &kfdefv1beta1.KfDef{}
+	if err := yaml.Unmarshal(configs, kfdef); err != nil {
+		return nil, &kfapis.KfError{
+			Code:    int(kfapis.INVALID_ARGUMENT),
+			Message: fmt.Sprintf("invalid config file format: %v", err),
+		}
+	}
+	return kfdef, nil
+}
+
+// Loads KfDef fild and returns KfDef in v1beta1. v1alpha1 will be converted to the new version.
+func LoadKfDefFromURI(configFile string) (*kfdefv1beta1.KfDef, error) {
+	if configFile == "" {
```

Comment:
So any KfDef config would be converted to the v1beta1 and then used?

Reply:
yes, after this call all KfDef should be in v1beta1.